File: code/programs/ruby/unix-tools/test/test_wc.rb
# frozen_string_literal: true

# test_wc.rb -- Tests for the Ruby wc tool
# ==========================================
#
# === What These Tests Verify ===
#
# These tests exercise the wc tool's CLI Builder integration and
# business logic. We test:
# - Line, word, and byte counting
# - Character counting (-m)
# - Maximum line length (-L)
# - Individual flag selection (-l, -w, -c)
# - Default output (lines + words + bytes)
# - Right-justified formatting
# - Total line for multiple files
# - The count_stream, format_counts, and compute_width functions directly

require "simplecov"
SimpleCov.start do
  add_filter "/test/"
  enable_coverage :branch
end

require "minitest/autorun"
require "stringio"
require "tempfile"
require "coding_adventures_cli_builder"

# Load the wc_tool module so we can test the business logic functions.
require_relative "../wc_tool"

# ---------------------------------------------------------------------------
# Helper module: shared spec path and parse method for wc tests
# ---------------------------------------------------------------------------

module WcTestHelper
  WC_TEST_SPEC = File.join(File.dirname(__FILE__), "..", "wc.json")

  def parse_wc_argv(argv)
    CodingAdventures::CliBuilder::Parser.new(WC_TEST_SPEC, ["wc"] + argv).parse
  end

  # Create a temporary file with given content, yield its path, then clean up.
  def with_tempfile(content)
    f = Tempfile.new("wc_test")
    f.write(content)
    f.close
    yield f.path
  ensure
    f&.unlink
  end
end

# ===========================================================================
# Test: CLI Builder integration
# ===========================================================================

class TestWcCliIntegration < Minitest::Test
  include WcTestHelper

  def test_no_flags_returns_parse_result
    result = parse_wc_argv([])
    assert_kind_of CodingAdventures::CliBuilder::ParseResult, result
  end

  def test_help_returns_help_result
    result = parse_wc_argv(["--help"])
    assert_kind_of CodingAdventures::CliBuilder::HelpResult, result
    assert_includes result.text, "wc"
  end

  def test_version_returns_version_result
    result = parse_wc_argv(["--version"])
    assert_kind_of CodingAdventures::CliBuilder::VersionResult, result
    assert_equal "1.0.0", result.version
  end

  def test_lines_flag
    result = parse_wc_argv(["-l"])
    assert result.flags["lines"]
  end

  def test_words_flag
    result = parse_wc_argv(["-w"])
    assert result.flags["words"]
  end

  def test_bytes_flag
    result = parse_wc_argv(["-c"])
    assert result.flags["bytes"]
  end

  def test_chars_flag
    result = parse_wc_argv(["-m"])
    assert result.flags["chars"]
  end

  def test_max_line_length_flag
    result = parse_wc_argv(["-L"])
    assert result.flags["max_line_length"]
  end
end

# ===========================================================================
# Test: count_stream function
# ===========================================================================

class TestWcCountStream < Minitest::Test
  def test_simple_text
    io = StringIO.new("hello world\n")
    counts = count_stream(io, "test.txt")
    assert_equal 1, counts.lines
    assert_equal 2, counts.words
    assert_equal 12, counts.bytes
    assert_equal 12, counts.chars
    assert_equal "test.txt", counts.name
  end

  def test_empty_input
    io = StringIO.new("")
    counts = count_stream(io, nil)
    assert_equal 0, counts.lines
    assert_equal 0, counts.words
    assert_equal 0, counts.bytes
    assert_nil counts.name
  end

  def test_multiple_lines
    io = StringIO.new("line one\nline two\nline three\n")
    counts = count_stream(io, "multi.txt")
    assert_equal 3, counts.lines
    assert_equal 6, counts.words
  end

  def test_no_trailing_newline
    # A file without a trailing newline: "hello" has 0 newline characters.
    io = StringIO.new("hello")
    counts = count_stream(io, "no_nl.txt")
    assert_equal 0, counts.lines
    assert_equal 1, counts.words
    assert_equal 5, counts.bytes
  end

  def test_max_line_length
    io = StringIO.new("short\na longer line\nhi\n")
    counts = count_stream(io, "test.txt")
    assert_equal 13, counts.max_line_length  # "a longer line" = 13 chars
  end

  def test_blank_lines
    io = StringIO.new("\n\n\n")
    counts = count_stream(io, "blanks.txt")
    assert_equal 3, counts.lines
    assert_equal 0, counts.words
  end

  def test_only_whitespace
    io = StringIO.new("   \n  \n")
    counts = count_stream(io, "ws.txt")
    assert_equal 2, counts.lines
    assert_equal 0, counts.words
  end
end

# ===========================================================================
# Test: format_counts function
# ===========================================================================

class TestWcFormatCounts < Minitest::Test
  def test_default_format
    counts = FileCounts.new(10, 42, 280, 280, 15, "file.txt")
    # Default mode: lines, words, bytes
    output = format_counts(counts, {}, 3)
    assert_includes output, " 10"
    assert_includes output, " 42"
    assert_includes output, "280"
    assert_includes output, "file.txt"
  end

  def test_lines_only
    counts = FileCounts.new(10, 42, 280, 280, 15, "file.txt")
    output = format_counts(counts, { "lines" => true }, 3)
    assert_includes output, " 10"
    refute_includes output, " 42"
    refute_includes output, "280"
    assert_includes output, "file.txt"
  end

  def test_words_only
    counts = FileCounts.new(10, 42, 280, 280, 15, "file.txt")
    output = format_counts(counts, { "words" => true }, 3)
    refute_includes output, " 10"
    assert_includes output, " 42"
    assert_includes output, "file.txt"
  end

  def test_bytes_only
    counts = FileCounts.new(10, 42, 280, 280, 15, "file.txt")
    output = format_counts(counts, { "bytes" => true }, 3)
    assert_includes output, "280"
    assert_includes output, "file.txt"
  end

  def test_chars_only
    counts = FileCounts.new(10, 42, 280, 250, 15, "file.txt")
    output = format_counts(counts, { "chars" => true }, 3)
    assert_includes output, "250"
    assert_includes output, "file.txt"
  end

  def test_max_line_length_only
    counts = FileCounts.new(10, 42, 280, 280, 15, "file.txt")
    output = format_counts(counts, { "max_line_length" => true }, 3)
    assert_includes output, " 15"
    assert_includes output, "file.txt"
  end

  def test_no_name
    counts = FileCounts.new(10, 42, 280, 280, 15, nil)
    output = format_counts(counts, {}, 3)
    refute_includes output, "nil"
  end
end

# ===========================================================================
# Test: compute_width function
# ===========================================================================

class TestWcComputeWidth < Minitest::Test
  def test_single_digit_counts
    counts = [FileCounts.new(5, 8, 9, 9, 3, "a.txt")]
    width = compute_width(counts, {})
    assert_equal 1, width
  end

  def test_multi_digit_counts
    counts = [FileCounts.new(100, 5000, 30000, 30000, 80, "big.txt")]
    width = compute_width(counts, {})
    assert_equal 5, width  # 30000 has 5 digits
  end

  def test_lines_only_width
    counts = [FileCounts.new(100, 5000, 30000, 30000, 80, "big.txt")]
    width = compute_width(counts, { "lines" => true })
    assert_equal 3, width  # 100 has 3 digits
  end
end

# ===========================================================================
# Test: Main function integration
# ===========================================================================

class TestWcMainFunction < Minitest::Test
  include WcTestHelper

  def test_main_reads_file
    with_tempfile("hello world\n") do |path|
      old_argv = ARGV.dup
      ARGV.replace([path])
      output = capture_io { wc_main }[0]
      # Should contain line count (1), word count (2), byte count (12)
      assert_match(/1/, output)
      assert_match(/2/, output)
      assert_match(/12/, output)
    ensure
      ARGV.replace(old_argv)
    end
  end

  def test_main_multiple_files_shows_total
    with_tempfile("aaa\n") do |path1|
      with_tempfile("bbb\n") do |path2|
        old_argv = ARGV.dup
        ARGV.replace([path1, path2])
        output = capture_io { wc_main }[0]
        assert_includes output, "total"
      ensure
        ARGV.replace(old_argv)
      end
    end
  end

  def test_main_nonexistent_file_prints_error
    old_argv = ARGV.dup
    ARGV.replace(["/nonexistent/file/xyz"])
    _stdout, stderr = capture_io { wc_main }
    assert_includes stderr, "No such file or directory"
  ensure
    ARGV.replace(old_argv)
  end

  def test_main_help_exits_zero
    old_argv = ARGV.dup
    ARGV.replace(["--help"])
    err = assert_raises(SystemExit) do
      capture_io { wc_main }
    end
    assert_equal 0, err.status
  ensure
    ARGV.replace(old_argv)
  end

  def test_main_version_exits_zero
    old_argv = ARGV.dup
    ARGV.replace(["--version"])
    err = assert_raises(SystemExit) do
      capture_io { wc_main }
    end
    assert_equal 0, err.status
  ensure
    ARGV.replace(old_argv)
  end

  def test_main_lines_only
    with_tempfile("alpha\nbeta\ngamma\n") do |path|
      old_argv = ARGV.dup
      ARGV.replace(["-l", path])
      output = capture_io { wc_main }[0]
      assert_match(/3/, output)
    ensure
      ARGV.replace(old_argv)
    end
  end

  def test_main_words_only
    with_tempfile("one two three\n") do |path|
      old_argv = ARGV.dup
      ARGV.replace(["-w", path])
      output = capture_io { wc_main }[0]
      assert_match(/3/, output)
    ensure
      ARGV.replace(old_argv)
    end
  end
end

File: code/programs/python/unix-tools/tests/test_wc.py
"""Tests for the wc tool.

=== What These Tests Verify ===

These tests exercise the wc implementation, including:

1. Count accuracy (lines, words, bytes, characters)
2. Flag parsing (-l, -w, -c, -m, -L)
3. Default behavior (show lines + words + bytes)
4. Column alignment and formatting
5. Total line for multiple files
6. CLI Builder integration (--help, --version)
7. The count_content and format_counts functions directly
"""

from __future__ import annotations

import sys
import tempfile
from pathlib import Path
from typing import Any

import pytest

# The spec file lives alongside the main script.
SPEC_FILE = str(Path(__file__).parent.parent / "wc.json")


# ---------------------------------------------------------------------------
# Helper: import cli_builder and parse argv
# ---------------------------------------------------------------------------


def parse_argv(argv: list[str]) -> Any:  # noqa: ANN401
    """Parse an argv list against the wc spec and return the result."""
    from cli_builder import Parser

    return Parser(SPEC_FILE, argv).parse()


# ---------------------------------------------------------------------------
# Test: Default behavior
# ---------------------------------------------------------------------------


class TestDefaultBehavior:
    """When invoked with no flags, wc should default to showing
    lines + words + bytes."""

    def test_no_flags_returns_parse_result(self) -> None:
        from cli_builder import ParseResult

        result = parse_argv(["wc"])
        assert isinstance(result, ParseResult)

    def test_no_specific_flags_set(self) -> None:
        from cli_builder import ParseResult

        result = parse_argv(["wc"])
        assert isinstance(result, ParseResult)
        # No specific counting flags should be set
        assert result.flags.get("lines") is not True
        assert result.flags.get("words") is not True
        assert result.flags.get("bytes") is not True

    def test_file_argument_parsed(self) -> None:
        from cli_builder import ParseResult

        result = parse_argv(["wc", "myfile.txt"])
        assert isinstance(result, ParseResult)
        files = result.arguments.get("files", [])
        assert "myfile.txt" in (files if isinstance(files, list) else [files])


# ---------------------------------------------------------------------------
# Test: Individual flags
# ---------------------------------------------------------------------------


class TestFlags:
    """Test that each flag is properly parsed."""

    def test_lines_flag_short(self) -> None:
        from cli_builder import ParseResult

        result = parse_argv(["wc", "-l"])
        assert isinstance(result, ParseResult)
        assert result.flags.get("lines") is True

    def test_lines_flag_long(self) -> None:
        from cli_builder import ParseResult

        result = parse_argv(["wc", "--lines"])
        assert isinstance(result, ParseResult)
        assert result.flags.get("lines") is True

    def test_words_flag_short(self) -> None:
        from cli_builder import ParseResult

        result = parse_argv(["wc", "-w"])
        assert isinstance(result, ParseResult)
        assert result.flags.get("words") is True

    def test_bytes_flag_short(self) -> None:
        from cli_builder import ParseResult

        result = parse_argv(["wc", "-c"])
        assert isinstance(result, ParseResult)
        assert result.flags.get("bytes") is True

    def test_chars_flag_short(self) -> None:
        from cli_builder import ParseResult

        result = parse_argv(["wc", "-m"])
        assert isinstance(result, ParseResult)
        assert result.flags.get("chars") is True

    def test_max_line_length_flag(self) -> None:
        from cli_builder import ParseResult

        result = parse_argv(["wc", "-L"])
        assert isinstance(result, ParseResult)
        assert result.flags.get("max_line_length") is True


# ---------------------------------------------------------------------------
# Test: count_content function
# ---------------------------------------------------------------------------


class TestCountContent:
    """Test the count_content function directly."""

    def test_empty_content(self) -> None:
        sys.path.insert(0, str(Path(__file__).parent.parent))
        from wc_tool import count_content

        counts = count_content("", b"")
        assert counts["lines"] == 0
        assert counts["words"] == 0
        assert counts["bytes"] == 0
        assert counts["chars"] == 0

    def test_single_line(self) -> None:
        sys.path.insert(0, str(Path(__file__).parent.parent))
        from wc_tool import count_content

        content = "hello world\n"
        raw = content.encode("utf-8")
        counts = count_content(content, raw)
        assert counts["lines"] == 1
        assert counts["words"] == 2
        assert counts["bytes"] == 12
        assert counts["chars"] == 12

    def test_multiple_lines(self) -> None:
        sys.path.insert(0, str(Path(__file__).parent.parent))
        from wc_tool import count_content

        content = "hello\nworld\n"
        raw = content.encode("utf-8")
        counts = count_content(content, raw)
        assert counts["lines"] == 2
        assert counts["words"] == 2
        assert counts["bytes"] == 12

    def test_no_trailing_newline(self) -> None:
        """A file without a trailing newline should still count words."""
        sys.path.insert(0, str(Path(__file__).parent.parent))
        from wc_tool import count_content

        content = "hello world"
        raw = content.encode("utf-8")
        counts = count_content(content, raw)
        assert counts["lines"] == 0  # No newline = 0 lines
        assert counts["words"] == 2
        assert counts["bytes"] == 11

    def test_max_line_length(self) -> None:
        sys.path.insert(0, str(Path(__file__).parent.parent))
        from wc_tool import count_content

        content = "short\nthis is a longer line\nhi\n"
        raw = content.encode("utf-8")
        counts = count_content(content, raw)
        assert counts["max_line_length"] == len("this is a longer line")

    def test_only_whitespace(self) -> None:
        sys.path.insert(0, str(Path(__file__).parent.parent))
        from wc_tool import count_content

        content = "   \n\n  \n"
        raw = content.encode("utf-8")
        counts = count_content(content, raw)
        assert counts["lines"] == 3
        assert counts["words"] == 0

    def test_multibyte_characters(self) -> None:
        """UTF-8 multibyte chars: bytes != chars."""
        sys.path.insert(0, str(Path(__file__).parent.parent))
        from wc_tool import count_content

        content = "cafe\u0301\n"  # e + combining accent
        raw = content.encode("utf-8")
        counts = count_content(content, raw)
        assert counts["chars"] == 6  # c, a, f, e, combining accent, \n
        assert counts["bytes"] == len(raw)  # More bytes due to UTF-8


# ---------------------------------------------------------------------------
# Test: format_counts function
# ---------------------------------------------------------------------------


class TestFormatCounts:
    """Test the format_counts function directly."""

    def test_all_counts(self) -> None:
        sys.path.insert(0, str(Path(__file__).parent.parent))
        from wc_tool import format_counts

        counts = {"lines": 5, "words": 12, "bytes": 68, "chars": 68, "max_line_length": 20}
        result = format_counts(
            counts,
            show_lines=True,
            show_words=True,
            show_bytes=True,
            show_chars=False,
            show_max_line_length=False,
            width=3,
            filename="test.txt",
        )
        assert "  5" in result
        assert " 12" in result
        assert " 68" in result
        assert "test.txt" in result

    def test_lines_only(self) -> None:
        sys.path.insert(0, str(Path(__file__).parent.parent))
        from wc_tool import format_counts

        counts = {"lines": 42, "words": 0, "bytes": 0, "chars": 0, "max_line_length": 0}
        result = format_counts(
            counts,
            show_lines=True,
            show_words=False,
            show_bytes=False,
            show_chars=False,
            show_max_line_length=False,
            width=2,
            filename=None,
        )
        assert result.strip() == "42"

    def test_no_filename(self) -> None:
        sys.path.insert(0, str(Path(__file__).parent.parent))
        from wc_tool import format_counts

        counts = {"lines": 1, "words": 2, "bytes": 3, "chars": 3, "max_line_length": 5}
        result = format_counts(
            counts,
            show_lines=True,
            show_words=True,
            show_bytes=True,
            show_chars=False,
            show_max_line_length=False,
            width=1,
            filename=None,
        )
        # Should not have a trailing filename
        assert result == "1 2 3"

    def test_right_alignment(self) -> None:
        sys.path.insert(0, str(Path(__file__).parent.parent))
        from wc_tool import format_counts

        counts = {"lines": 1, "words": 2, "bytes": 3, "chars": 3, "max_line_length": 0}
        result = format_counts(
            counts,
            show_lines=True,
            show_words=True,
            show_bytes=True,
            show_chars=False,
            show_max_line_length=False,
            width=5,
        )
        assert "    1" in result
        assert "    2" in result
        assert "    3" in result


# ---------------------------------------------------------------------------
# Test: Full integration with real files
# ---------------------------------------------------------------------------


class TestFileIntegration:
    """Test wc with actual temporary files."""

    def test_count_real_file(self, capsys: pytest.CaptureFixture[str]) -> None:
        sys.path.insert(0, str(Path(__file__).parent.parent))
        from wc_tool import main

        with tempfile.NamedTemporaryFile(mode="w", suffix=".txt", delete=False) as f:
            f.write("hello world\ngoodbye world\n")
            f.flush()
            tmpfile = f.name

        old_argv = sys.argv
        try:
            sys.argv = ["wc", tmpfile]
            main()
        finally:
            sys.argv = old_argv
            Path(tmpfile).unlink()

        captured = capsys.readouterr()
        # Should have 2 lines, 4 words, and the correct byte count
        assert "2" in captured.out
        assert "4" in captured.out


# ---------------------------------------------------------------------------
# Test: --help and --version
# ---------------------------------------------------------------------------


class TestHelpFlag:
    """``--help`` should return a HelpResult."""

    def test_help_returns_help_result(self) -> None:
        from cli_builder import HelpResult

        result = parse_argv(["wc", "--help"])
        assert isinstance(result, HelpResult)

    def test_help_text_contains_program_name(self) -> None:
        from cli_builder import HelpResult

        result = parse_argv(["wc", "--help"])
        assert isinstance(result, HelpResult)
        assert "wc" in result.text


class TestVersionFlag:
    """``--version`` should return a VersionResult with "1.0.0"."""

    def test_version_returns_version_result(self) -> None:
        from cli_builder import VersionResult

        result = parse_argv(["wc", "--version"])
        assert isinstance(result, VersionResult)

    def test_version_string(self) -> None:
        from cli_builder import VersionResult

        result = parse_argv(["wc", "--version"])
        assert isinstance(result, VersionResult)
        assert result.version == "1.0.0"
